R/test-server.R

```diff
@@ -3,6 +3,27 @@ isModuleServer <- function(x) {
   is.function(x) && names(formals(x))[1] == "id"
 }
 
+#' @noRd
+isAppDir <- function(path) {
+
+  if (file.exists(file.path(path, "app.R")))
+    return(TRUE)
+
+  if (file.exists(file.path(path, "server.R"))
+      && file.exists(file.path(path, "ui.R")))
+    return(TRUE)
+
+  FALSE
+}
+
+#' @noRd
+findEnclosingApp <- function(path = ".") {
+  rprojroot::find_root(
```

Comment:
I think we should do this without taking a dependency on rprojroot, since the task is relatively small.

Reply:
OK, shall I copy the relevant code from rprojroot and include it in Shiny, or write new, rprojroot-inspired code?

I decided to depend on rprojroot mostly because it encapsulates subtleties related to functioning properly on Windows.

Reply:
What kind of issues are there with Windows?

Reply:
There are no issues per se; there are only decisions about which conventions to use, such as which value to use consistently for the `winslash` argument of `normalizePath()` or how many parents to traverse before erroring. The benefit I saw to depending on rprojroot was less that it provided some algorithm that was difficult to reproduce, and more that it encapsulates a standard approach to this mundane problem, with limitations and capabilities that are already universally understood. One potential end-user benefit of this kind of reuse is that related error messages might be easier to Google.

That said, I understand that adding dependencies carries its own tradeoffs, and an argument for reuse based on convention adoption is relatively weak. It's straightforward for me to write new code in Shiny that follows the `winslash` and `.MAX_DEPTH` conventions established by rprojroot but is otherwise more focused on our particular use-case. Shall I do that?

Reply:
Since we're just using this internally, can we just do something simple like use `normalizePath()` and then call `dirname()` until we hit the top level?

Reply:
I'll try it out, and I think for cross-platform root detection we can borrow this from rprojroot: https://github.com/r-lib/rprojroot/blob/master/R/root.R#L67-L73